include/aspect/melt.h

```diff
@@ -0,0 +1,279 @@
+ */
+
+
+#ifndef __aspect__melt_h
+#define __aspect__melt_h
+
+#include <aspect/simulator_access.h>
+#include <aspect/global.h>
+#include <aspect/assembly.h>
+#include <aspect/material_model/interface.h>
+#include <aspect/fluid_pressure_boundary_conditions/interface.h>
+
+namespace aspect
+{
+  using namespace dealii;
+
+  namespace MaterialModel
+  {
+    template <int dim>
+    class MeltOutputs : public AdditionalMaterialOutputs<dim>
+    {
+      public:
+        MeltOutputs (const unsigned int n_points,
+                     const unsigned int /*n_comp*/)
+        {
+          compaction_viscosities.resize(n_points);
+          fluid_viscosities.resize(n_points);
+          permeabilities.resize(n_points);
+          fluid_densities.resize(n_points);
+          fluid_density_gradients.resize(n_points, Tensor<1,dim>());
+        }
+
+        /**
+         * Compaction viscosity values $\xi$ at the given positions.
+         * This parameter describes the resistance of the solid matrix
+         * in a two-phase simulation to dilation and compaction.
+         */
+        std::vector<double> compaction_viscosities;
+
+        /**
+         * Fluid (melt) viscosity values $\eta_f$ at the given positions.
+         */
+        std::vector<double> fluid_viscosities;
+
+        /**
+         * Permeability values $k$ at the given positions.
+         */
+        std::vector<double> permeabilities;
+
+        /**
+         * Fluid (melt) density values $\rho_f$ at the given positions.
+         */
+        std::vector<double> fluid_densities;
+
+        /**
+         * An approximation for the fluid (melt) density gradients
+         * $\nabla \rho_f$ at the given positions. These values are
+         * required for compressible models to describe volume changes
+         * of melt in dependence of pressure, temperature etc.
+         */
+        std::vector<Tensor<1,dim> > fluid_density_gradients;
+
+        void average (const MaterialAveraging::AveragingOperation operation,
+                      const FullMatrix<double>  &projection_matrix,
+                      const FullMatrix<double>  &expansion_matrix)
+        {
+          average_property (operation, projection_matrix, expansion_matrix,
+                            compaction_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            permeabilities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_densities);
+
+          // the fluid density gradients are unfortunately stored in reverse
+          // indexing. it's also not quite clear whether these should
+          // really be averaged, so avoid this for now
+        }
+    };
+
+    /**
+     * Base model for material models that implement a melt fraction function.
```

Comment:
I think it would be clearer if you said "Base CLASS for material models that implement..."